frontend/catalyst/jax_extras/tracing.py

```diff
@@ -399,9 +556,423 @@ def make_jaxpr_f(*args, **kwargs):
+        outputs: Expanded list of outputs of the program
+
+    Returns:
+        New OutputType containing no InDBIdx references to the specified argument.
+    """
+    assert len(out_type) == len(outputs), "Outputs must be expanded"
+    x_in_idx = input_idx
+    x_out_idx = tracer_index([*consts, *inputs][x_in_idx], outputs)
+
+    out_type2 = []
+    for i, ((aval, k), t) in enumerate(zip(out_type, outputs)):
+        assert hasattr(t, "aval"), "Outputs are expected to be Jax tracers or Jaxpr variables"
+        if isinstance(aval, DShapedArray):
+            shape2 = []
+            for d in aval.shape:
+                if isinstance(d, InDBIdx) and d.val == x_in_idx:
+                    assert x_out_idx is not None, (
+                        "Target tracer does not exist in the outputs "
+                        "(see force_implicit_indbidx=True)"
+                    )
+                    assert x_out_idx < i, "Target tracer is not available for OutDBIdx"
+                    shape2.append(OutDBIdx(x_out_idx))
+                else:
+                    shape2.append(d)
+            aval2 = aval.update(shape=tuple(shape2))
+        else:
+            aval2 = copy(aval)
+        out_type2.append((aval2, k))
+    return out_type2
+
+
+class DynshapePrimitive(JaxprPrimitive):
+    """Primitive containing nested Jaxpr programs accepting and returning Jax values with shapes
+    containing dynamic dimensions."""
+
+    def bind(self, *args, **params):
+        """Bind the Jax primitive into a Jaxpr program. This method is called during both the
+        tracing of a Python program and during the evaluation of a Jaxpr program.
+
+        In contrast to other Jax primitives, this one accepts expanded arguments (shapes of
+        inputs might contain tracers), deduces the Jax ``InputType``, passes it to the abstract
+        evaluation callback and correctly interprets the returned ``OutputType``. This means that it
+        correctly processes the InDBIdx/OutDBIdx residing in the OutputType by creating resulting
+        tracers with the right dynamic dimensions. For the details, refer to the definition of the
+        ``InputType`` and ``OutputType`` of Jax.
+
+        Args:
+            *args: arguments of this bind primitive. Must be Jax tracers (when tracing a Python
+                   program) or Jaxpr variables (when evaluating Jaxpr program). Dynamic dimensions
+                   are supported.
+            **params: Any valid Python variables used as parameters for this primitive.
+
+        Returns:
+            Output Python tracers or Jaxpr variables, with the dynamic dimensions set in accordance
+            with the the output type calculated by the ``abstract_eval``.
+
+        """
+        # Note: We do not return `out_type` along with the results in order to match the Jax bind
+        # API.  but it would be good to return it, because otherwise we lose the information about
+        # explicintess.
```

Comment:
```suggestion
        # Note: We do not return `out_type` along with the results in order to match the Jax bind
        # API, but it would be good to return it, because otherwise we lose the information about
        # explicitness.